plugins/woocommerce/includes/class-woocommerce.php

```diff
@@ -665,6 +676,11 @@ public function includes() {
 			$this->frontend_includes();
 		}
 
+		if ( $this->is_rest_api_request() && ! $this->is_store_api_request() ) {
+			// Include tracks classes for use in REST API.
+			$this->tracks_includes();
```

Comment:
This adds Woo tracking to _any_ REST request in the context of WordPress right (except for Woo Store API requests). That still seems problematic because there could be other WordPress plugins (like Woo) that do Store API requests.

Another likely problem with the current condition here is that the Store API _might_ be used in the context of the Admin (eg previews) so are we sure tracking should be disabled in that case?

I don't think the above questions/concerns need to be handled in this in initial PR, but I do think we should be looking at ways to only enable tracking on _explicit_ routes it should be enabled on. Given the increasing overlap between REST Requests and the admin context this might need some sort of route level registration.

Reply:
Any Store API request is by definition a customer request, even if done by a merchant. Any merchant tracking done on that path shouldn't trigger. For the example, the goal of enabling tracking in rest requests is that if you save a settings page (via REST API), any tracking attached to that request should trigger (like we do for Local Pickup settings page). If you trigger a Store API requests, there aren't any merchant tracking events attached to it.

So if you're in the editor for Products page, a Store API request is triggered to get all products, that request should not have any merchant tracking in it, and if it does, it won't load.

If you modify said block and save the page, that request can have merchant tracking in it and it will be loaded.

I asked for Store API to be excluded because:
- It's customer only, we shouldn't risk having merchant related things loaded in it.
- Loading all the tracks classes in Store API adds unnecessary load.

>  That still seems problematic because there could be other WordPress plugins (like Woo) that do Store API requests.

I'm not sure I understand this concern?

Reply:
> I'm not sure I understand this concern?

It's not with your explanation, thanks Nadir.

Reply:
Actually, no what I meant is that there are other plugins that do register REST routes and might be doing REST requests that shouldn't have our tracks loaded for? The current implementation is all REST requests (except Store API).

Reply:
That's a good point, the current check applies to all rest requests, this means this logic will load for any rest request (be it meant for merchants or customers).

For now, the main concern of loading it for all REST API requests is performance, this is no difference from what it was previously (since 7.8). We should look at building that list of paths indeed, it should include all of `/wc` namespaced routes, except `wc/store`, and also includes a handpicked list of `/wp` namespaced requests (page saves, settings saves, and possibly any Gutenberg related endpoints).

Reply:
I agree that we should either move to an allowed list of endpoints that can use tracking and/or guarantee that tracking methods are inert when tracking is disabled.

If we go with the former, we'll need to do a full audit of tracking done in REST endpoints to make sure we don't miss any that may have been added since https://github.com/woocommerce/woocommerce/pull/37796.  This is a bit problematic because these are not easy to trace as the tracks calls may not be within the endpoints themselves, but within models or other service classes.

Instead, it might be more feasible to make sure that only REST API requests coming from admin users (those opted into tracking) can be tracked.

Calling of a method or script on tracking should not trigger any actual tracks or enqueueing of scripts unless tracking has been opted in by the requesting user (the admin).

>For now, the main concern of loading it for all REST API requests is performance, this is no difference from what it was previously (since 7.8).

@senadir can you expand on the performance concerns?  This might dictate which path we take.